gherkin/java/src/main/java/io/cucumber/gherkin/TokenMatcher.java

```diff
@@ -166,29 +173,51 @@ private boolean match_DocStringSeparator(Token token, String separator, boolean
                 indentToRemove = 0;
             }
 
-            setTokenMatched(token, TokenType.DocStringSeparator, mediaType, separator, null, null);
+            setTokenMatched(token, TokenType.DocStringSeparator, mediaType, separator, null, null, null);
             return true;
         }
         return false;
     }
 
     @Override
     public boolean match_StepLine(Token token) {
-        List<String> keywords = currentDialect.getStepKeywords();
-        for (String keyword : keywords) {
-            if (token.line.startsWith(keyword)) {
-                String stepText = token.line.getRestTrimmed(keyword.length());
-                setTokenMatched(token, TokenType.StepLine, stepText, keyword, null, null);
-                return true;
+        Map<StepKeywordType, List<String>> keywords = currentDialect.getStepKeywordsByType();
+
+        String keyword = null;
+        StepKeywordType keywordType = StepKeywordType.UNKNOWN;
+        for (Entry<StepKeywordType,List<String>> entry : keywords.entrySet()) {
+            List<String> translations = entry.getValue();
+
+            if (keyword != null) { // we have a keyword already
+                if (translations.contains(keyword)) {
+                    keywordType = StepKeywordType.UNKNOWN;
```

Comment:
`keywordType` is already set in the `else` branch. Why would it have to be set?

Reply:
(discussion continued in the other comment)